internal/resources/google/sql_database_instance.go

```diff
@@ -0,0 +1,314 @@
+package google
+
+import (
+	"github.com/infracost/infracost/internal/resources"
+	"github.com/infracost/infracost/internal/schema"
+
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/shopspring/decimal"
+)
+
+type SQLDatabaseInstance struct {
+	Address              string
+	DiskSize             int64
+	UseIPV4              bool
+	ReplicaConfiguration string
+	Tier                 string
+	AvailabilityType     string
+	Region               string
+	DatabaseVersion      string
+	DiskType             string
+	BackupStorageGB      *float64 `infracost_usage:"backup_storage_gb"`
+}
+
+var SQLDatabaseInstanceUsageSchema = []*schema.UsageItem{{Key: "backup_storage_gb", ValueType: schema.Float64, DefaultValue: 0}}
+
+func (r *SQLDatabaseInstance) PopulateUsage(u *schema.UsageData) {
+	resources.PopulateArgsWithUsage(r, u)
+}
+
+func (r *SQLDatabaseInstance) BuildResource() *schema.Resource {
+	var resource *schema.Resource
+
+	replica := false
+	if r.ReplicaConfiguration != "" {
+		replica = true
+	}
+
+	resource = r.sqlDatabaseInstanceCostComponents(false, r.Address)
+	if replica {
+		resource.SubResources = append(resource.SubResources, r.sqlDatabaseInstanceCostComponents(true, "Replica"))
+	}
+
+	return resource
+}
+
+type SQLInstanceDBType int
+
+const (
+	MySQL SQLInstanceDBType = iota
+	PostgreSQL
+	SQLServer
+)
+
+func (r *SQLDatabaseInstance) dbType() SQLInstanceDBType {
+	return r.sqlInstanceDBVersionToDBType()
+}
+
+func (r *SQLDatabaseInstance) sqlDatabaseInstanceCostComponents(replica bool, name string) *schema.Resource {
+	var costComponents []*schema.CostComponent
+	tier := r.Tier
+
+	if r.AvailabilityType == "" || replica {
+		r.AvailabilityType = "ZONAL"
+	}
+
+	if r.DiskType == "" {
+		r.DiskType = "PD_SSD"
+	}
+
+	if r.DiskSize == 0 {
+		r.DiskSize = 10
+	}
+
+	if r.sqlInstanceTierToResourceGroup() != "" && r.dbType() != SQLServer {
+		costComponents = append(costComponents, r.sharedSQLInstance())
+	} else if r.sqlInstanceTierToResourceGroup() == "" && strings.Contains(tier, "db-custom-") {
+		cpu, _ := strconv.ParseInt(strings.Split(tier, "-")[2], 10, 32)
```

Comment:
Should we check for errors here? I'm assuming `[2]` won't cause any issues because we validate there's 2 `-`s in the string above, but the `ParseInt` could cause an error.